server/src/main/java/org/elasticsearch/action/bulk/SimulateBulkRequest.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.bulk;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * This extends BulkRequest with support for providing substitute pipeline definitions. In a user request, the pipeline substitutions
+ * will look something like this:
+ *
+ *   "pipeline_substitutions": {
+ *     "my-pipeline-1": {
+ *       "processors": [
+ *         {
+ *           "set": {
+ *             "field": "my-new-boolean-field",
+ *             "value": true
+ *           }
+ *         }
+ *       ]
+ *     },
+ *     "my-pipeline-2": {
+ *       "processors": [
+ *         {
+ *           "set": {
+ *             "field": "my-new-boolean-field",
+ *             "value": true
+ *           },
+ *           "rename": {
+ *               "field": "old_field",
+ *               "target_field": "new field"
+ *           }
+ *         }
+ *       ]
+ *     }
+ *   }
+ *
+ *   The pipelineSubstitutions Map held by this class is intended to be the result of XContentHelper.convertToMap(). The top-level keys
+ *   are the pipelineIds ("my-pipeline-1" and "my-pipeline-2" in the example above). The values are the Maps of "processors" to the List of
+ *   processor definitions.
+ */
+public class SimulateBulkRequest extends BulkRequest {
+    private Map<String, Map<String, Object>> pipelineSubstitutions = Map.of();
```

Comment:
I think if we add a constructor for the `pipelineSubstitutions` then we can make this `final`, think it's worth it?

Reply:
Oh you're right -- I think this came from a very early iteration where I was modifying the BulkRequest class directly.